src/util.c

```diff
@@ -808,4 +808,158 @@ bool check_is_ip_addr_valid_local(int ss_family, char *ip_to_check)
+            addr = &((struct sockaddr_in6 *)ifa->ifa_addr)->sin6_addr;
+        } else {
+            continue;
+        }
+
+        inet_ntop(family, addr, ip, sizeof(ip));
+
+        if (strcmp(ip, target_ip) == 0) {
+            strncpy(iface_name, ifa->ifa_name, IFNAMSIZ - 1);
+            iface_name[IFNAMSIZ - 1] = '\0';
+            freeifaddrs(ifaddr);
+            return 0;
+        }
+    }
+
+    freeifaddrs(ifaddr);
+    return 1;  // No match found
+}
+
+/**
+ * Bind a socket to a local address and port, optionally using a specific client address.
+ *
+ * @param sockfd        The socket file descriptor to bind.
+ * @param client_port   The local port number to bind the socket to.
+ * @param sc            Pointer to the ntttcp_stream_client structure containing socket configuration,
+ *                      including address family and optional client address.
+ *
+ * This function sets up the local address structure based on the address family (IPv4 or IPv6)
+ * and attempts to bind the socket to the specified port. If a client address is provided and
+ * valid, it will bind to that address; otherwise, it binds using the default interface.
+ *
+ * Note: If two interfaces share the same subnet, bind() alone does not determine the outgoing
+ * interface. In such cases, SO_BINDTODEVICE should be used to bind to a specific device.
+ *
+ * @return 0 on success, or a negative value if bind() fails.
+ */
+
+int ntttcp_bind_socket(int sockfd, int client_port, struct ntttcp_stream_client *sc)
+{
+	char *log = NULL;
+	struct sockaddr_storage local_addr; /* for local address */
+	socklen_t local_addr_size = sizeof(local_addr); /* local address size */
+	char if_name[IFNAMSIZ] = {'\0'};
+	int ret = 0; /* hold function return value */
+
+	memset(&local_addr, 0, sizeof(local_addr));
+
+	if (sc->domain == AF_INET) {
+		(*(struct sockaddr_in *)&local_addr).sin_family = AF_INET; /* local_addrs[i].ss_family = AF_INET; */
+		(*(struct sockaddr_in *)&local_addr).sin_port = htons(client_port);
+
+		if(sc->use_client_address)
+		{
+			if (inet_pton(AF_INET, sc->client_address,  &((*(struct sockaddr_in *)&local_addr).sin_addr)) <= 0) {
+				ASPRINTF(&log, "Invalid IPv4 address or Address %s not supported", sc->client_address);
+				/* Allow to go through the default interface */
+			}
+		}
+	} else {
+		(*(struct sockaddr_in6 *)&local_addr).sin6_family = AF_INET6; /* local_addrs[i].ss_family = AF_INET6; */
+		(*(struct sockaddr_in6 *)&local_addr).sin6_port = htons(client_port);
+
+		if(sc->use_client_address)
+		{
+			if (inet_pton(AF_INET6, sc->client_address,  &((*(struct sockaddr_in6 *)&local_addr).sin6_addr)) <= 0) {
+				ASPRINTF(&log, "Invalid IPv6 address or Address %s not supported", sc->client_address);
+				/* Allow to go through the default interface */
+			}
+		}
+	}
+
+	if ((ret = bind(sockfd, (struct sockaddr *)&local_addr, local_addr_size)) < 0) {
+		ASPRINTF(&log,
+				"failed to bind socket[%d] to a local : [%s]. errno = %d. Ignored",
```

Comment:
Can we keep the original log format that is "failed to bind socket[%d] to a local port: [%s:%d]. errno = %d. Ignored"?